module/VuFind/src/VuFind/Db/Service/TagService.php

```diff
@@ -778,4 +780,399 @@ public function getStatistics(bool $extended = false): array
         }
         return $stats;
     }
+
+    /**
+     * Create a Tags entity object.
+     *
+     * @return Tags
+     */
+    public function createTags(): Tags
+    {
+        $class = $this->getEntityClass(Tags::class);
+        return new $class();
+    }
+
+    /**
+     * Get the row associated with a specific tag string.
+     *
+     * @param string $tag       Tag to look up.
+     * @param bool   $create    Should we create the row if it does not exist?
+     * @param bool   $firstOnly Should we return the first matching row (true)
+     * or the entire result set (in case of multiple matches)?
+     *
+     * @return mixed Matching row/result set if found or created, null otherwise.
+     */
+    public function getByText($tag, $create = true, $firstOnly = true)
+    {
+        $dql = 'SELECT t '
+            . 'FROM ' . $this->getEntityClass(Tags::class) . ' t '
+            . 'WHERE ' . ($this->caseSensitive ? 't.tag = :tag' : 'LOWER(t.tag) = LOWER(:tag) ');
+        $parameters = compact('tag');
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        $result =  $query->getResult();
+        if (count($result) == 0 && $create) {
+            $row = $this->createTags()
+                ->setTag($this->caseSensitive ? $tag : mb_strtolower($tag, 'UTF8'));
+            try {
+                $this->persistEntity($row);
+            } catch (\Exception $e) {
+                $this->logError('Could not save tag: ' . $e->getMessage());
+                return false;
+            }
+            return $firstOnly ? $row : [$row];
+        }
+        return $firstOnly ? current($result) : $result;
+    }
+
+    /**
+     * Get a list of tags based on a sort method ($sort) and a where clause.
+     *
+     * @param string $sort  Sort/search parameter
+     * @param int    $limit Maximum number of tags (default = 100,
+     *                      < 1 = no limit)
+     * @param string $where Extra code to modify query (null for none)
```

Comment:
I wonder if it would be better to make this a little more generic -- instead of using "magic strings" and a switch statement, what if we instead make $where an array of clauses, and replace $text with a $parameters argument? Then we could replace the switch with something like:

```
if (!empty($where)) {
    $dql .= ' WHERE ' . implode(' AND ', $where);
}
```

And then we can just move the setting of parameters and WHERE clauses into the calling code. Right now, it seems like each type of call to this method is only used in one place, so I don't think it's bad to move the details to the caller... but if we wanted to make it more reusable, we could create public wrapper methods in this class -- e.g. getNonListTags() or getTagListMatchingText().